modules/fx/arrow-fx/src/main/kotlin/arrow/fx/Queue.kt

```diff
@@ -198,6 +208,61 @@ class Queue<F, A> private constructor(private val capacity: Int, private val ref
         ifShutdown = ::identity
       )
     }
+
+  private sealed class State<F, out A> {
+    abstract fun size(): Kind<F, Int>
+
+    internal data class Deficit<F, A>(val takers: IQueue<Promise<F, A>>, val AP: Applicative<F>, val shutdownHook: Kind<F, Unit>) : State<F, A>() {
+      override fun size(): Kind<F, Int> = AP.just(-takers.length())
+    }
+
+    internal data class Surplus<F, A>(val queue: IQueue<A>, val putters: IQueue<Tuple2<A, Promise<F, Unit>>>, val AP: Applicative<F>, val shutdownHook: Kind<F, Unit>) : State<F, A>() {
+      override fun size(): Kind<F, Int> = AP.just(queue.length() + putters.length())
+    }
+
+    internal data class Shutdown<F>(val AE: ApplicativeError<F, Throwable>) : State<F, Nothing>() {
+      override fun size(): Kind<F, Int> = AE.raiseError(QueueShutdown)
+    }
+  }
+
+  private sealed class SurplusStrategy<F, A> {
+    abstract fun handleSurplus(p: Promise<F, Unit>, surplus: State.Surplus<F, A>, a: A): Tuple2<Kind<F, Unit>, State<F, A>>
+
+    internal data class Blocking<F, A>(val capacity: Int, val AP: Applicative<F>) : SurplusStrategy<F, A>() {
```

Comment:
All these `internal` modifiers are unnecessary since the sealed class is `private` all its sub types also are also `private`.

```suggestion
    data class Blocking<F, A>(val capacity: Int, val AP: Applicative<F>) : SurplusStrategy<F, A>() {
```

Reply:
We should probably name this the same as the constructor like the others.

```suggestion
    internal data class Bounded<F, A>(val capacity: Int, val AP: Applicative<F>) : SurplusStrategy<F, A>() {
```